src/ADAL.PCL.Desktop/CryptographyHelper.cs

```diff
@@ -54,33 +54,66 @@ public byte[] SignWithCertificate(string message, X509Certificate2 certificate)
-            RSACryptoServiceProvider rsa = x509Key.GetAsymmetricAlgorithm(SecurityAlgorithms.RsaSha256Signature, true) as RSACryptoServiceProvider;
+            byte[] messageBytes = Encoding.UTF8.GetBytes(message);
+            var x509Key = new X509AsymmetricSecurityKey(certificate);
+
+            RSA rsa = x509Key.GetAsymmetricAlgorithm(SecurityAlgorithms.RsaSha256Signature, true) as RSA;
 
             RSACryptoServiceProvider newRsa = null;
             try
             {
-                newRsa = GetCryptoProviderForSha256(rsa);
-                using (SHA256Cng sha = new SHA256Cng())
+                if (rsa is RSACryptoServiceProvider cspRsa)
+                {
+                    // For .NET 4.6 and below we get the old RSACryptoServiceProvider implementation as the default.
+                    // Try and get an instance of RSACryptoServiceProvider which supports SHA256
+                    newRsa = GetCryptoProviderForSha256(cspRsa);
+                }
+                else
                 {
-                    return newRsa.SignData(Encoding.UTF8.GetBytes(message), sha);
+                    // For .NET Framework 4.7 and onwards the RSACng implementation is the default.
+                    // Since we're targeting .NET Framework 4.5, we cannot actually use this type as it was
+                    // only introduced with .NET Framework 4.6.
+                    // Instead we try and create an RSACryptoServiceProvider based on the private key from the
+                    // certificate.
+                    newRsa = GetCryptoProviderForSha256(certificate);
+                }
+
+                using (var sha = new SHA256Cng())
+                {
+                    return newRsa.SignData(messageBytes, sha);
                 }
             }
             finally
             {
+                // We only want to dispose of the 'newRsa' instance if it is a *different instance*
+                // from the original one that was used to create it.
                 if (newRsa != null && !ReferenceEquals(rsa, newRsa))
                 {
                     newRsa.Dispose();
                 }
             }
         }
 
+        /// <summary>
+        /// Create a <see cref="RSACryptoServiceProvider"/> using the private key from the given <see cref="X509Certificate2"/>.
+        /// </summary>
+        /// <param name="certificate">Certificate including private key with which to initialize the <see cref="RSACryptoServiceProvider"/> with</param>
+        /// <returns><see cref="RSACryptoServiceProvider"/> initialized with private key from <paramref name="certificate"/></returns>
+        private static RSACryptoServiceProvider GetCryptoProviderForSha256(X509Certificate2 certificate)
+        {
+            var privateKeyXmlParams = certificate.PrivateKey.ToXmlString(true);
+            var rsa = new RSACryptoServiceProvider();
+            rsa.FromXmlString(privateKeyXmlParams);
```

Comment:
Basically here we're allowing the new `RSACryptoServiceProvider` to rebuild the private key from the various parameters.
The other method (below) creates a `RSACSP` instance and passes in the key container and number.

These should be equivalent.